docs/articles/persistence/custom-persistence-provider.md

```diff
@@ -0,0 +1,467 @@
+* The `Task` must be completed with a failure if any of the persistent messages could not be replayed.
+
+The `toSequenceNr` is the lowest of what was returned by `ReadHighestSequenceNrAsync` and what the user specified as the recovery `Recovery` parameter. This does imply that this call is always preceded by reading the highest sequence number for the given `persistenceId`.
+
+This call is NOT protected with a circuit-breaker because it may take a long time to replay all events. The plugin implementation itself must protect against an unresponsive backend store and make sure that the returned `Task` is completed with success or failure within reasonable time. It is not allowed to ignore completing the `Task`.
+
+##### Code Sample
+
+`ByPersistenceIdSql` in this example refers to this SQL query statement:
+
+[!code-csharp[ByPersistenceIdSql](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ByPersistenceIdSql "ByPersistenceIdSql SQL Statement")]
+
+SQLite code example:
+
+[!code-csharp[ReplayMessagesAsync](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ReplayMessagesAsync "ReplayMessagesAsync method implementation")]
+
+#### ReadHighestSequenceNrAsync
+
+```c#
+Task<long> ReadHighestSequenceNrAsync(
+    string persistenceId, 
+    long fromSequenceNr)
+```
+
+* **`persistenceId`**: Persistent actor identifier
+* **`fromSequenceNr`**: Hint where to start searching for the highest sequence number. When a persistent actor is recovering this `fromSequenceNr` will the sequence number of the used snapshot, or `0L` if no snapshot is used.
+
+This call is protected with a circuit-breaker.
+
+This method should asynchronously reads the highest stored sequence number for provided `persistenceId`. The persistent actor will use the highest sequence number after recovery as the starting point when persisting new events. This sequence number is also used as `toSequenceNr` in subsequent calls to `ReplayMessagesAsync` unless the user has specified a lower `toSequenceNr`. Journal must maintain the highest sequence number and never decrease it.
+
+Please also note that requests for the highest sequence number may be made concurrently to writes executing for the same `persistenceId`, in particular it is possible that a restarting actor tries to recover before its outstanding writes have completed.
+
+`HighestSequenceNrSql` in this example refers to this SQL query statement:
+
+[!code-csharp[HighestSequenceNrSql](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=HighestSequenceNrSql "HighestSequenceNrSql SQL Statement")]
+
+SQLite code example:
+
+[!code-csharp[ReadHighestSequenceNrAsync](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ReadHighestSequenceNrAsync "ReadHighestSequenceNrAsync method implementation")]
+
+#### WriteMessagesAsync
+
+```c#
```

Comment:
Do we leave this as it is or make use of the range selection though the line number may change in the future?